pageserver/src/walingest.rs

```diff
@@ -347,16 +432,38 @@ impl WalIngest {
                         // we could peek into the message and only pause if it contains
                         // a particular string, for example, but this is enough for now.
                         failpoint_support::sleep_millis_async!("wal-ingest-logical-message-sleep");
+                        special_treatment_check!(needs none);
                     } else if let Some(path) = prefix.strip_prefix("neon-file:") {
                         modification.put_file(path, message, ctx).await?;
+                    } else {
+                        special_treatment_check!(needs none);
                     }
+                } else {
+                    special_treatment_check!(unknown record type, pg_version, lsn, decoded);
                 }
             }
-            _x => {
-                // TODO: should probably log & fail here instead of blindly
-                // doing something without understanding the protocol
-            }
-        }
+            pg_constants::RM_STANDBY_ID => special_treatment_check!(needs none),
+
+            // All of these are handled by the generic ingest_decoded_block function
+            pg_constants::RM_BTREE_ID => special_treatment_check!(needs none),
+            pg_constants::RM_HASH_ID => special_treatment_check!(needs none),
+            pg_constants::RM_GIN_ID => special_treatment_check!(needs none),
+            pg_constants::RM_GIST_ID => special_treatment_check!(needs none),
+            pg_constants::RM_SEQ_ID => special_treatment_check!(needs none),
+            pg_constants::RM_SPGIST_ID => special_treatment_check!(needs none),
+            pg_constants::RM_BRIN_ID => special_treatment_check!(needs none),
+            pg_constants::RM_GENERIC_ID => special_treatment_check!(needs none),
+
+            // We don't support the commit-ts tracking in neon. No harm if we see
+            // these records though.
+            pg_constants::RM_COMMIT_TS_ID => special_treatment_check!(needs none),
+
+            // These are related to logical replication. I don't know if we should
+            // do something with them. @knizhnik?
+            pg_constants::RM_REPLORIGIN_ID => special_treatment_check!(needs none),
+
+            _x => special_treatment_check!(unknown record type, pg_version, lsn, decoded),
```

Comment:
If I understand correctly, here is the only place when the macro stops being so useful as we have possibly many `_x`.. But since we this should be really rare perhaps we could afford a static mutex guarded hashset of `_x`?

Or if this is already an exhaustive match, we could mention it in a comment (really difficult to eyeball is it or not).